### ISSUE TITLE
Wrong poisson distribution sample at large lambda

### ISSUE THREAD

**Issue by kalradivyanshu** [jstat/jstat#243]
for large lambda (like 604800) the sample from poisson distribution is wrong.
jStat.poisson.sample(604800)
>696
jStat.poisson.sample(604800)
>741
jStat.poisson.sample(604800)
>737
jStat.poisson.sample(604800)
>735
jStat.poisson.sample(604800)
>771
jStat.poisson.sample(604800)
751
jStat.poisson.sample(604800)
>728
jStat.poisson.sample(604800)
>738
jStat.poisson.sample(604800)
>767
jStat.poisson.sample(604800)
>691

Compared to python's scipy:

>>> poisson(604800).rvs()
604888
>>> poisson(604800).rvs()
604955
>>> poisson(604800).rvs()
604769
>>> poisson(604800).rvs()
604747
>>> poisson.rvs(604800)
605129
>>> poisson.rvs(604800)
604286
>>> poisson.rvs(604800)
604801
>>> poisson.rvs(604800)
603856
>>> poisson.rvs(604800)
604193

**Comment by trevnorris** [comment]
Can you confirm if it returns the correct value when the sample size is small?

**Comment by kalradivyanshu** [comment]
So at after around 750 lambda, it keeps returning samples around 750 only. I ran a test, taking 10000 samples from jstat and 10000 from scipy and comparing the average. Here is the code:
jStat:
`var jStat = require('jStat').jStat;

function getAverage(lambda) {
    let sumSamples = 0;
    for(let i = 0; i < 10000; i++)
        sumSamples += jStat.poisson.sample(lambda);
    return sumSamples/10000;
}

for(let i = 0; i < 10000; i+=10) {
    console.log(i, getAverage(i));
}`

Scipy:

`from scipy.stats import poisson
import numpy as np

i = 0
while i < 10000:
    print(i, np.mean(poisson(i).rvs(10000)))
    i += 10
`

Here are the results:

![image](https://user-images.githubusercontent.com/12642750/59084379-a0b5cc00-8918-11e9-991e-9808a65b8a4f.png)

If we zoom in on xaxis : [0, 1000]:

![image](https://user-images.githubusercontent.com/12642750/59084405-b925e680-8918-11e9-940f-8874e0e602ad.png)

As you can see it is correct for lambda <= 750, after that it just keeps repeating samples as if the lambda was 750.

Thanks.

**Comment by kalradivyanshu** [comment]
So I can see the problem, jStat uses Math.exp(-lambda) to sample from poisson distribution, and at 750, that number becomes so small that it rounds it off to zero:

```
lambda: Math.exp(-lambda)
700: 9.85967654375977e-305
710: 4.47628622567513e-309
720: 2.0322308024e-313
730: 9.226315e-318
740: 4.2e-322
750: 0
760: 0
770: 0
```

Not sure how to fix it I am looking into it, any help would be appreciated.

**Comment by trevnorris** [comment]
@kalradivyanshu excellent info. Here's an alternative function:
```js
function sample(l) {
  let p = 0;
  let k = 0;
  do {
    k++;
    p += Math.log(Math.random());
  } while (-p < l);
  return k - 1;
}
```
And averages taken from it:
```
1000: 994.2  
2000: 2000.92
3000: 2998   
4000: 3998.16
5000: 4983.32
6000: 5988.48
7000: 6981.36
8000: 8005.64
604800: 604705.72
```
I have no idea if doing the math that way is technically correct, but it does keep the values within a manageable range and returns the expected averages.

**Comment by trevnorris** [comment]
Wrote this node.js script to check on the distribution of values using the above `sample` function.
```js
'use strict';

const VAL = 500;
const ITER = 300;

const { stdev, poisson } = require(process.cwd() + '/dist/jstat');
const s1_arr = [];
const s2_arr = [];

function sample(l) {
  let p = 0;
  let k = 0;
  do {
    k++;
    p += Math.log(Math.random());
  } while (-p < l);
  return k - 1;
}

function gen_sample_array(iter, val, fn) {
  const a = [];
  for (let i = 0; i < iter; i++)
    a.push(fn(val));
  return a;
}

console.log(stdev(gen_sample_array(ITER, VAL, poisson.sample)).toFixed(4),
            stdev(gen_sample_array(ITER, VAL, sample)).toFixed(4));

for (let i = 0; i < ITER; i++) {
  s1_arr.push(stdev(gen_sample_array(ITER, VAL, poisson.sample)));
  s2_arr.push(stdev(gen_sample_array(ITER, VAL, sample)));
}

console.log(stdev(s1_arr).toFixed(4), stdev(s2_arr).toFixed(4));
```
Here's some example output. The first row is the std dev of `500` values. Second row is the std dev of `500` std devs of the `sample` functions.
```
21.7977 21.7423
0.9422 0.8553
```
Not sure if this shows the alternative function also works.

**Comment by kalradivyanshu** [comment]
I went into the source code of numpy to see how they are handling it, since numpy is very fast when generating poisson samples, and found this [function](https://github.com/numpy/numpy/blob/5bd7c2f41ff28cecc9f594e3d5e69de046d15b35/numpy/random/src/distributions/distributions.c#L668), So I ported it to javascript, and it is very fast:

```
function loggam(x) {
    let x0, x2, xp, gl, gl0;
    let k, n;
  
    let a = [8.333333333333333e-02, -2.777777777777778e-03,
            7.936507936507937e-04, -5.952380952380952e-04,
            8.417508417508418e-04, -1.917526917526918e-03,
            6.410256410256410e-03, -2.955065359477124e-02,
            1.796443723688307e-01, -1.39243221690590e+00];
    x0 = x;
    n = 0;
    if ((x == 1.0) || (x == 2.0)) {
        return 0.0;
    } else if (x <= 7.0) {
        n = Math.floor(7 - x);
        x0 = x + n;
    }
    x2 = 1.0 / (x0 * x0);
    xp = 2 * 3.14159265358979323846264338327950288;
    gl0 = a[9];
    for (k = 8; k >= 0; k--) {
        gl0 *= x2;
        gl0 += a[k];
    }
    gl = gl0 / x0 + 0.5 * Math.log(xp) + (x0 - 0.5) * Math.log(x0) - x0;
    if (x <= 7.0) {
        for (k = 1; k <= n; k++) {
            gl -= Math.log(x0 - 1.0);
            x0 -= 1.0;
        }
    }
    return gl;
}
function sample(lam) {
    let k;
    let U, V, slam, loglam, a, b, invalpha, vr, us;

    slam = Math.sqrt(lam);
    loglam = Math.log(lam);
    b = 0.931 + 2.53 * slam;
    a = -0.059 + 0.02483 * b;
    invalpha = 1.1239 + 1.1328 / (b - 3.4);
    vr = 0.9277 - 3.6224 / (b - 2);

    while (1) {
        U = Math.random() - 0.5;
        V = Math.random();
        us = 0.5 - Math.abs(U);
        k = Math.floor((2 * a / us + b) * U + lam + 0.43);
        if ((us >= 0.07) && (V <= vr)) {
            return k;
        }
        if ((k < 0) || ((us < 0.013) && (V > us))) {
            continue;
        }
        /* log(V) == log(0.0) ok here */
        /* if U==0.0 so that us==0.0, log is ok since always returns */
        if ((Math.log(V) + Math.log(invalpha) - Math.log(a / (us * us) + b)) <= (-lam + k * loglam - loggam(k + 1))) {
            return k;
        }
    }
}

function sample_2(l) {
    let p = 0;
    let k = 0;
    do {
      k++;
      p += Math.log(Math.random());
    } while (-p < l);
    return k - 1;
}

function getAverage2(lambda) {
    let sumSamples = 0;
    for(let i = 0; i < 10000; i++)
        sumSamples += sample_2(lambda);
    return sumSamples/10000;
}

function getAverage(lambda) {
    let sumSamples = 0;
    for(let i = 0; i < 10000; i++)
        sumSamples += sample(lambda);
    return sumSamples/10000;
}

console.time("sample_numpy");
console.log(60000, getAverage(60000));
console.timeEnd("sample_numpy");

console.time("sample");
console.log(60000, getAverage2(60000));
console.timeEnd("sample");
```

where getAverage2 uses the function @trevnorris suggested, and getAverage uses the numpy port function:

Here is the timing data:
```
60000 60000.3125
sample_numpy: 39.010ms
60000 59996.5793
sample: 13249.343ms
```
So it is almost 350x faster.

Found [this](https://pdfs.semanticscholar.org/00f1/8468dd53e4e3f0c28a5d504f8cd10376a673.pdf) paper that explains the algorithm used. I am trying to understand it. The parameters used in the algorithm is with the assumption that lambda >= 10. So jStat should have that, if lambda is less than 10, it calls the current function and if not call this function.

I will love to add it to library, but I will need some guidance as to how to structure the code. Thanks.

**Comment by kalradivyanshu** [comment]
Any updates on this?

**Comment by paulmarx** [comment]
Fuck you!!!

Don’t send me emails!!!

I will sew you if i‘ll receive another one!!!

**Comment by trevnorris** [comment]
@kalradivyanshu I'd suggest replacing/implementing the new code in the way that seems most reasonable to you. After you post the PR I'll walk you through any changes, or make code suggestions. I think that'd be the fastest way to go.

**Comment by kalradivyanshu** [comment]
I have added the code and opened a pull request, it would be great, if you could give some feedback, thanks. https://github.com/jstat/jstat/pull/246

**Comment by trevnorris** [comment]
This is fixed by a22a7f2.
